### PR TITLE
[Client] Asyncio Client, Sync gRPC Server

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

* Supercedes #15275
* Uses a similar dedicated thread for processing the gRPC `request_queue` as [LogsServicer](https://github.com/ray-project/ray/blob/b08b2c5103c634c680de31b237b2bfcceb9bc150/python/ray/util/client/server/logservicer.py#L55)

## NOTE: 
This is **not a perfect asyncio implementation**. It is possible for the async return to be delayed by _up to **MAX_BLOCKING_OPERATION_TIME_S**_ because there is only one loop per client. This should not functionally be a big issue because in such a case, the Client would be blocked on a `ray.get` or ray.wait` call. 

This is better than the previous asyncio approach (in #15275) because in that approach, asyncio awaits would be correctly scheduled and processed, but `ray.get`s from **different clients** would block each other. This implementation fully maintains the existing isolation between clients. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/client/common.py [hunk 1]
```diff
@@ -60,7 +62,51 @@ def __del__(self):
 
 
 class ClientObjectRef(ClientBaseRef):
-    pass
+    def __await__(self):
+        return self.as_future().__await__()
+
+    def as_future(self) -> asyncio.Future:
```
Comment (simon-mo) [inline python/ray/util/client/common.py:68]: can we adopt the new `ref.future() -> concurrent.futures.Future` API? see object_ref.pyx

**Review by simon-mo — APPROVED** [review]

File: python/ray/util/client/common.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/util/client/common.py:69]: Yep, ptal

Comment (simon-mo) [reply python/ray/util/client/common.py:69]: yeah lg!

File: python/ray/util/client/server/dataservicer.py [hunk 2]
```diff
@@ -34,7 +35,33 @@ def Datapath(self, request_iterator, context):
...
+                try:
+                    for req in request_iterator:
+                        request_queue.put(req)
+                except grpc.RpcError as e:
+                    logger.debug("closing dataservicer reader thread "
+                                 f"grpc error reading request_iterator: {e}")
+                finally:
+                    request_queue.put(None)
+
+            queue_filler_thread = Thread(target=fill_queue, daemon=True)
```
Comment (msilvestriUSF) [inline python/ray/util/client/server/dataservicer.py:53]: @ijrsvt 
Check out thread resource exhaustion like in [logservicer.py](https://github.com/ray-project/ray/blob/57c0bd9912caffc4cbb8a895784242663031a156/python/ray/util/client/server/logservicer.py#L93) and implement similar logic here.

This may change the factor of `CLIENT_SERVER_MAX_THREADS` to be `1/3`, but probably not.

File: python/ray/tests/test_client_library_integration.py [hunk 3]
```diff
@@ -22,5 +22,23 @@ def test_rllib_integration(ray_start_regular_shared):
...
+        def hello(request):
+            return "hello"
+
+        serve.create_backend("my_backend", hello, config={"num_replicas": 1})
+        serve.create_endpoint(
+            "my_endpoint", backend="my_backend", route="/hello")
+        handle = serve.get_handle("my_endpoint")
+        assert ray.get(handle.remote()) == "hello"
+        assert await handle.remote() == "hello"
+
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_library_integration.py:41]: Awesome, thanks for adding that!

File: python/ray/util/client/dataclient.py [hunk 4]
```diff
@@ -35,6 +34,11 @@ def __init__(self, channel: "grpc._channel.Channel", client_id: str,
         self.ready_data: Dict[int, Any] = {}
         self.cv = threading.Condition()
         self.lock = threading.RLock()
+
+        # NOTE: Dictionary insertion is guaranteed to complete before lookup
+        # and/or removal because of synchronization via the request_queue.
+        self.asyncio_waiting_data: Dict[int, Callable[
+            [ray_client_pb2.DataResponse], None]] = {}
```
Comment (AmeerHajAli) [inline python/ray/util/client/dataclient.py:41]: should we give Dict[int, Callable[[ray_client_pb2.DataResponse], None]] a type name at the top of the file?

File: python/ray/util/client/dataclient.py [hunk 5]
```diff
@@ -66,9 +70,16 @@ def _data_main(self) -> None:
                     # This is not being waited for.
                     logger.debug(f"Got unawaited response {response}")
                     continue
-                with self.cv:
-                    self.ready_data[response.req_id] = response
-                    self.cv.notify_all()
+                if response.req_id in self.asyncio_waiting_data:
+                    cb = self.asyncio_waiting_data.pop(response.req_id)
```
Comment (AmeerHajAli) [inline python/ray/util/client/dataclient.py:74]: what is cb?

File: python/ray/util/client/dataclient.py [hunk 6]
```diff
@@ -35,6 +34,11 @@ def __init__(self, channel: "grpc._channel.Channel", client_id: str,
         self.ready_data: Dict[int, Any] = {}
         self.cv = threading.Condition()
```
Comment (AmeerHajAli) [inline python/ray/util/client/dataclient.py:35]: what is self.cv ? can we use another name for this parameter?

File: python/ray/util/client/server/dataservicer.py [hunk 7]
```diff
@@ -19,6 +20,20 @@
 logger = logging.getLogger(__name__)
 
 
+def fill_queue(grpc_input_generator, output_queue) -> None:
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:23]: type annotation

File: python/ray/util/client/server/dataservicer.py [hunk 8]
```diff
@@ -19,6 +20,20 @@
...
+    """
+    Pushes incoming requests to a shared output_queue.
+    """
+    try:
+        for req in grpc_input_generator:
+            output_queue.put(req)
+    except grpc.RpcError as e:
+        logger.debug("closing dataservicer reader thread "
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:31]:
```suggestion
        logger.exception("closing dataservicer reader thread "
```

File: python/ray/util/client/server/dataservicer.py [hunk 9]
```diff
@@ -19,6 +20,20 @@
...
+                     f"grpc error reading request_iterator: {e}")
+    finally:
+        output_queue.put(None)
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:34]: why is this necessary?

File: python/ray/util/client/server/dataservicer.py [hunk 10]
```diff
@@ -33,23 +48,47 @@ def Datapath(self, request_iterator, context):
             logger.error("Client connecting with no client_id")
             return
         logger.debug(f"New data connection from client {client_id}: ")
+        if not self._init(client_id, context):
+            return
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:52]: why moving this out is OK?
why we no longer need to give the RESOURCE_EXHAAUSTED status code?

File: python/ray/util/client/server/dataservicer.py [hunk 11]
```diff
@@ -33,23 +48,47 @@ def Datapath(self, request_iterator, context):
...
+                            if get_resp is None:
+                                # The response for this request will be sent
+                                # when the object is ready.
+                                continue
+                        else:
+                            get_resp = self.basic_service._get_object(
+                                req.get, client_id)
                         resp = ray_client_pb2.DataResponse(get=get_resp)
                     elif req_type == "put":
                         put_resp = self.basic_service._put_object(
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:94]: Can we split these into two function calls one for asyncio and one for non asyncio?
I am trying to think of the best way to make this more readable. Maybe there are other ways to do it and I will refer here to your wisdom.

File: python/ray/util/client/server/dataservicer.py [hunk 12]
```diff
@@ -82,7 +121,7 @@ def Datapath(self, request_iterator, context):
         finally:
             logger.debug(f"Lost data connection from client {client_id}")
             self.basic_service.release_all(client_id)
-
+            queue_filler_thread.join()
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:124]: is this always guaranteed to succeed?

File: python/ray/util/client/server/dataservicer.py [hunk 13]
```diff
@@ -96,7 +135,11 @@ def Datapath(self, request_iterator, context):
                         logger.debug("Shutting down ray.")
                         ray.shutdown()
 
-    def _init(self, req_init, client_id):
+    def _init(self, client_id, context):
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:138]: type annotation :D

File: python/ray/util/client/server/dataservicer.py [hunk 14]
```diff
@@ -33,23 +48,47 @@ def Datapath(self, request_iterator, context):
...
-                    if resp is None:
-                        context.set_code(grpc.StatusCode.RESOURCE_EXHAUSTED)
-                        return
+                    resp_init = self.basic_service.Init(req.init)
+                    resp = ray_client_pb2.DataResponse(init=resp_init, )
                     logger.debug(f"Accepted data connection from {client_id}. "
                                  f"Total clients: {self.num_clients}")
                     accepted_connection = True
                 else:
                     assert accepted_connection
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/dataservicer.py:79]: is this assert necessary?

File: python/ray/util/client/server/server.py [hunk 15]
```diff
@@ -250,6 +251,49 @@ def Terminate(self, req, context=None):
...
+            result_queue: queue.Queue,
+            context=None) -> Optional[ray_client_pb2.GetResponse]:
+        """Attempts to schedule a callback to push the GetResponse to the
+        main loop when the desired object is ready. If there is some failure
+        in scheduling, a GetResponse will be immediately returned.
+        """
+        if request.id not in self.object_refs[client_id]:
+            return ray_client_pb2.GetResponse(valid=False)
+        try:
+            assert request.asynchronous
```
Comment (AmeerHajAli) [inline python/ray/util/client/server/server.py:268]: is this assert necessary?

**Comment by AmeerHajAli** [comment]
hey @ijrsvt , I am not sure I understand the Note, can you please elaborate one the pros and cons of each approach?
Also CC @edoakes , I think it would be valuable if you review this.

File: python/ray/util/client/dataclient.py (hunk 4 above)
Comment (ijrsvt) [reply python/ray/util/client/dataclient.py:41]: I'll give the callback a typename

Comment (ijrsvt) [reply python/ray/util/client/dataclient.py:41]: Great idea!

File: python/ray/util/client/dataclient.py (hunk 5 above)
Comment (ijrsvt) [reply python/ray/util/client/dataclient.py:74]: callback

File: python/ray/util/client/dataclient.py (hunk 6 above)
Comment (ijrsvt) [reply python/ray/util/client/dataclient.py:35]: It is a condition variable. This is used pretty consistently throughout the Ray Client codebase.

File: python/ray/util/client/server/dataservicer.py (hunk 14 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:79]: Yes, because it ensures init was called first.

File: python/ray/util/client/server/dataservicer.py (hunk 10 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:52]: WDYM? if `_init()` returns False, it means initialization failed, so we just return from the function, which makes the server close the connection.

File: python/ray/util/client/server/dataservicer.py (hunk 9 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:34]: Because the iterator over the `output_queue` stops iterating when it sees a `None` value.

File: python/ray/util/client/server/dataservicer.py (hunk 11 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:94]: @AmeerHajAli we cannot split `Datapath` into two separate functions. It is a stub method used by gRPC. I think leaving as is makes the most sense. Let me add an assert or two to make it clear what the logic is.

File: python/ray/util/client/server/dataservicer.py (hunk 12 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:124]: If it does not succeed, then there is a problem. I will switch to a timeout, with an exception. This is most likely to cause a problem in tests, not actual uses. Actual uses are easier to cleanup because `ray stop` forcibly kills processes.

File: python/ray/util/client/server/server.py (hunk 15 above)
Comment (ijrsvt) [reply python/ray/util/client/server/server.py:268]: Not really.

File: python/ray/util/client/server/dataservicer.py (hunk 8 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py:31]: @AmeerHajAli Most places that catch `grpc.RpcError`s just Debug log. I would prefer to keep this the same as all those calls.

File: python/ray/util/client/server/dataservicer.py (hunk 11 above)
Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py]: I actually removed one level of nesting, because the `accepted_connection` boolean **should be set ** by the `self._init` function. The reason for this is that `accepted_connection` was used to ensure that we decrement `num_clients` if and only if we had incremented that. Now, if `num_clients` is **not incremented** (meaning we don't have resources), the function returns early.

File: python/ray/util/client/server/dataservicer.py (hunk 14 above)
Comment (AmeerHajAli) [reply python/ray/util/client/server/dataservicer.py]: but why do we need it? do we want to fail if accepted_connection is False? Shouldn't we instead check if it is true and if not raise an exception?

Comment (ijrsvt) [reply python/ray/util/client/server/dataservicer.py]: This has been removed as of my commit from yesterday.

**Review by edoakes — APPROVED** [review]
